packages/editor/src/utils/firebaseUtils.tsx

```diff
@@ -0,0 +1,163 @@
+          /**
+           * Save access token to local state
+           * settingsEffect propogates to local storage
+           */
+          set(settingsState, (prevState) => ({
+            ...prevState,
+            githubAccessToken:
+              credential.accessToken != undefined // For type safety, assign null if undefined
+                ? credential.accessToken
+                : null,
+          }));
+
+          toast.success(`Logged in as ${result.user.displayName}`);
+        }
+      })
+      .catch((error) => {
+        toast.error("Error logging in");
+      });
+  });
+
+export const signOutWrapper = () =>
+  useRecoilCallback(({ set }) => async () => {
+    signOut(authObject)
+      .then(() => {
+        set(settingsState, (prevState) => ({
+          ...prevState,
+          githubAccessToken: null,
+        }));
+        toast.success("Logged out");
+      })
+      .catch((error) => {
+        toast.error("Error logging out");
+      });
+  });
+
+// Firestore database utils
+export function createWorkspaceObject(
+  name: string,
+  lastModified: number,
+  id: string,
+  editorVersion: number,
+  saved: boolean,
+  substance: string,
+  style: string,
+  domain: string,
+): Workspace {
+  return {
+    metadata: {
+      name: name,
+      lastModified: lastModified,
+      id: id,
+      editorVersion: editorVersion,
+      forkedFromGist: null,
+      location: { kind: "stored", saved: saved },
+    },
+    files: {
+      substance: {
+        name: ".substance",
+        contents: substance,
+      },
+      style: {
+        name: ".style",
+        contents: style,
+      },
+      domain: {
+        name: ".domain",
+        contents: domain,
+      },
+    },
+  };
+}
+
+/*
+ * Creates hashmap of workspace ids to workspace objects from user's cloud
+ * storage. Used to populate savedFilesState
+ */
+export async function createSavedWorkspaceObject(userid: string) {
+  var loadedWorkspaces = {} as SavedWorkspaces;
+  const querySnapshot = await getDocs(collection(db, userid));
+
+  querySnapshot.forEach((doc) => {
+    var docData = doc.data();
+    loadedWorkspaces[doc.id] = createWorkspaceObject(
+      docData.name,
+      docData.lastModified,
+      docData.diagramId,
+      docData.editorVersion,
+      true,
+      docData.substance,
+      docData.style,
+      docData.domain,
+    );
+  });
+
+  return loadedWorkspaces;
+}
+
+export async function getDiagram(diagramId: string) {
+  if (authObject.currentUser != null) {
+    const fetchedDoc = await getDoc(
+      doc(db, authObject.currentUser.uid, diagramId),
+    );
+    if (fetchedDoc.exists()) {
+      const docData = fetchedDoc.data();
+      return createWorkspaceObject(
+        docData.name,
+        docData.lastModified,
+        docData.diagramId,
+        docData.editorVersion,
+        true,
+        docData.substance,
+        docData.style,
+        docData.domain,
+      );
+    } else {
+      toast.error("Could not fetch diagram");
```

Comment:
```suggestion
      toast.error(`Could not fetch diagram ${name} from cloud storage because it doesn't exist.`);
```

Reply:
Updated to error message: "Could not fetch diagram from cloud storage because it doesn't exist."
getDiagram would be called from loadWorkspace, where name isn't defined. This is called using diagramId as an identifier which isn't meaningful to the user (It's a randomly generated string)